xarray/core/duck_array_ops.py

```diff
@@ -158,17 +158,23 @@ def trapz(y, x, axis):
 )
 
 
-def asarray(data):
+def asarray(data, xp=np):
     return (
         data
         if (isinstance(data, dask_array_type) or hasattr(data, "__array_function__"))
-        else np.asarray(data)
+        else xp.asarray(data)
```

Comment:
I think we should instead update this so it detects duck arrays more reliably (I guess `cupy` doesn't implement `__array_function__`?).

Reply:
CuPy implements `__array_function__`.  I believe the issue here is how to handle int/floats/etc. When we start mixing numpy/cupy data types functions like `np.where(cupy-array > 0, numpy.array, cupy.array)`will fail. 

xarray ends up in the situation during: `[asarray(x) for x in scalars_or_arrays]` where `scalars_or_arrays` is a mixed list due to `int` not having `__array_function__`.  So [int, cupy.array] -> [numpy.array, cupy.array]

Reply:
thanks, that makes sense. I'd still like a way to make that work for every duck array without special casing, but I have no idea how that could be done.

Reply:
I totally agree. This quick WIP shows a viable solution for special casing cupy, but isn't satisfying.

In NEPs 35 and 37 (which are still in draft) there are diverging discussions on handling cases like this. The main point is changing `asarray` to take another array as a reference. 

NEP 35 for example would result in something like:

```python
np.asarray(scalar_value, like=cupy_or_pint_array)
```

In this case you would just pass another element from `scalars_or_arrays` to set the reference type. We may need some logic to work out which is the reference type if there are multiple other types though.

Reply:
cc @pentschev who may have insight into what to do here

Reply:
Unfortunately, I don't think there's much we can do with this piece of code at the moment, `__array_function__` still has the limitation of not being able to dispatch on array creation. As @jacobtomlinson 's suggested, the proposal from NEP-35 to add `like=` would be the appropriate solution IMO, but that is not yet implemented in NumPy. I've been (finally) working on that this week, but given this is a very sensitive change in NumPy's code I think it's going to take a while until it lands.

Dask is probably the library in which we worked the most to extend `__array_function__` to its limits, but we still encounter the same limitations of not being able to dispatch on creation. For the time being, I don't think there's a solution other than doing something like the code here is doing.

Reply:
For everyone's awareness, I've submitted https://github.com/numpy/numpy/pull/16935 to get started with the NEP-35 implementation I mentioned in the comment above.